webrender/src/device/gfx/device.rs

```diff
@@ -1414,9 +1414,35 @@ impl<B: hal::Backend> Device<B> {
             .programs
             .get_mut(&self.bound_program)
             .expect("Program not found.");
-        let mut cmd_buffer = self.command_pool[self.next_id].acquire_command_buffer();
-        unsafe { cmd_buffer.begin() };
         let descriptor_group = program.shader_kind.into();
+
+        let mut barriers: SmallVec<[hal::memory::Barrier<B>; RENDERER_TEXTURE_COUNT]> = SmallVec::new();
+        for index in &self.bound_textures {
+            let image = &self.images[index].core;
+            if let Some(barrier) = image.transit(
+                hal::image::Access::SHADER_READ,
+                hal::image::Layout::ShaderReadOnlyOptimal,
+                image.subresource_range.clone(),
+                None,
+            ) {
+                barriers.push(barrier)
+            }
+        };
+        if !barriers.is_empty() {
+            let cmd_buffer = self.command_pool[self.next_id].acquire_command_buffer();
```

Comment:
This isn't going to work - it's a separate command buffer that we aren't even submitting anywhere.
Even if we did - the barriers would happen before any other work we do, so if a resource needs to change its state throughout the frame (i.e. switch from being a render target to being a source texture), we'd not have that.

Reply:
It's not a separate command buffer, maybe you got confused by the function name `acquire_command_buffer` which has the same name as the `gfx-hal` function, but it's on our side and returns a mutable reference of a command buffer from a vector of command buffers:
https://github.com/szeged/webrender/blob/master/webrender/src/device/gfx/command.rs#L25
Later we submit those buffers for the current frame in `submit_to_gpu`.

Reply:
ok, good. The case I make in the review is still actual though - we can't make those transitions in the middle of a render pass

Reply:
This part of code only runs before we begin a render pass (right now we only have one render pass for each frame), so I don't think the case you mentioned would occur with our current implementation. However I can agree that on the longer run we need to end up with an implementation which you covered in the B section. In this case we will have more render passes, and most of the transitions would be a part of those.

Reply:
ok, let's not forget to talk about this on the call tomorrow